prow/crier/reporters/github/reporter.go

```diff
@@ -63,6 +63,7 @@ func (s *shardedLock) getLock(key simplePull) *sync.Mutex {
 	if _, exists := s.locks[key]; !exists {
 		s.locks[key] = &sync.Mutex{}
 	}
+	s.locks[key].Lock()
```

Comment:
I am not really fond of the approach, because it means that we have to hold the map lock (which is at the end of the day a global lock around all presubmit reporting) while waiting for lock, causing a lot of contention.

If we can not think of a way to avoid that, I'd prefer to keep this the way it is because:
* The race you described is generally not very likely to happen
* Even if it ends up happening, the actual race that this is preventing can only happen for failed jobs when they modify the github comment in parallel, discarding each others changes

Reply:
Thats a good point, I overlooked that. That problem can be significantly reduced by using a read-write mutex so that we only hold a global read lock when waiting for the individual locks. I thought it would eliminate the problem by limiting contention to the once per hour cleanup, but I just remembered there would still be contention when adding a key for a new PR.
Since the `sync.RWMutex` doesn't support atomic upgrade/downgrade, using a RWMutex also required a locking loop which isn't as straight forward as a `Lock` and deferred `Unlock`.

I agree that this race is too unlikely to occur to be worth the reduced performance and any extra complexity. The expected behavior if the race does somehow occur isn't critical either.
Thanks for the review, Alvaro!
/close